### DISCUSSION TITLE
Using dependencies in exception handlers

### DISCUSSION THREAD

**Discussion by septatrix** [fastapi/fastapi#8339] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from sqlalchemy.exc import SQLAlchemyError

app = FastAPI()

def handle_sqlalchemy(request: Request, error, user = Depends(get_current_user)):
    if user.is_admin:
        # return more verbose information which should not be shown to unprivileged users
    raise HTTPException(HTTPStatus.INTERNAL_SERVER_ERROR)

app.add_exception_handler(SQLAlchemyError, handle_sqlalchemy)
```

### Description

Is it possible to use dependency injection inside error handlers added via `app.add_exception_handler`?
There are often scenarios where different actions have to be taken depending on or using data from dependencies.

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

Python 3.10.5

### Additional Context

The following error is shown indicating that dependencies are not resolved: `AttributeError: 'Depends' object has no attribute 'is_admin'`

If there were a possibility to resolve dependencies manually that would also be sufficient (e.g. `app.resolve(Depends(...))`) however I do not know of any such functionality either.

**Comment by odiseo0** [comment, marked as answer]
`Depends` only works on FastAPI endpoints, you could use another approach using the `request` object

**Comment by TalalMahmoodChaudhry** [comment]
How to use request object?

What I tried was trying to pass in the DB client through the Exception but it doesn't work. Getting error `AttributeError: 'NoneType' object has no attribute 'request'` for below code. I think DB client is already closed by then maybe.
`

```
from fastapi import FastAPI
from azure.cosmos.aio import CosmosClient as CosmosClient

async def get_cosmos_async_client() -> CosmosClient:
    db = CosmosClient(cosmos_url, credential=cosmos_key)
    try:
        yield db
    finally:
        await db.close()
        
async def update_database_with_error(client: CosmosClient) -> Dict[str, Any]:
    database = client.get_database_client(DATABASE)
    container = database.get_container_client(CONTAINER_NAME)
    item = await container.upsert_item(body=<Some Dict>)
    return item
        
app = FastAPI()

@app.put('/update',)
async def update_data(data: Data, client=Depends(get_cosmos_async_client)):
   try:
       await update_database_with_data(data)
   except Exception:
       raise DBError("Failed to update", client)

async def handle_update_failed(request: Request, exc: DBError) -> JSONResponse:
    await update_database_with_error(exc.args[1])
    return JSONResponse(status_code=500, content={'error': exc.args[0]})

app.add_exception_handler(DBError, handle_update_failed)

**Comment by chrisPiemonte** [comment]
so no solution so far to this ?

**Reply by WilliamStam** [reply]
the first parameter on the exception handlers is the request object. but it doesnt really help much im afraid :( seems to run a lot earlier than when the other parts get added in. 

```
@app.exception_handler(Exception)
async def http_server_exception_handler(request, exc: Exception):
    error_type = type(exc).__name__
    error_message = str(exc)
    stack_trace = traceback.extract_tb(exc.__traceback__)
    print(request.scope)
```

so even if you have a dependency on a route that adds something to the request object, and the route triggers an error you dont get that "hydrated" request object. its a somewhat strange design choice i feel

other option might be to use middleware for it and have a try catch around the execution part
